crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -82,29 +85,32 @@ impl AuthorityStorePruner {
         }
         if !indirect_objects.is_empty() {
             let ref_count_update = indirect_objects
-                .into_iter()
+                .iter()
                 .map(|(digest, delta)| (digest, delta.to_le_bytes()));
             wb = wb.partial_merge_batch(&perpetual_db.indirect_move_objects, ref_count_update)?;
         }
+        wb = perpetual_db.set_highest_pruned_checkpoint(wb, checkpoint_number)?;
+
+        let _locks = objects_lock_table
+            .acquire_locks(indirect_objects.into_keys())
```

Comment:
If the number of `indirect_objects` here is too large then we may wait on some locks while holding onto others (which will block readers). One way to avoid that would be to process smaller number of tx effects in one write batch (to incur smaller number of indirect_objects that we might need to lock). Another option would be to just use `try_lock` with a sleep and retry (since pruning is not critical). We could even do something interesting like use `try_lock` for some fixed number of iterations before timing out and falling back to `acquire_lock` here (The overall goal is to avoid trying to take a lock on bunch of objects simultaneously). A reader biased rw-lock might also be an option. I wonder if @andll has some thoughts on this.

Reply:
@sadhansood very good points, but given that pruner will run pretty rarely, it should be okay to wait and see with this. it would be good however to monitor how long we spend waiting to acquire locks here.

Reply:
I think one thing that can happen in that case is that we are doing a lot of try_lock and then end up locking/unlocking some locks if other fail to lock. So I would not do this optimization before we can actually measure if it makes things better. One thing we can always do is adding `monitored_scope` to see how much time this place take